app/src/main/java/org/schabi/newpipe/info_list/InfoItemDialog.java

```diff
@@ -1,54 +1,296 @@
+            this(activity, context, fragment, infoItem, true);
+        }
+
+        /**
+         * <p>Create an instance of this Builder.</p>
+         * <p>If {@code addDefaultEntriesAutomatically} is set to {@code true},
+         * some default entries are added to the top and bottom of the dialog.</p>
+         * The dialog has the following structure:
+         * <pre>
+         *     + - - - - - - - - - - - - - - - - - - - - - -+
+         *     | ENQUEUE                                    |
+         *     | ENQUEUE_HERE                               |
+         *     | START_ON_BACKGROUND                        |
+         *     | START_ON_POPUP                             |
+         *     + - - - - - - - - - - - - - - - - - - - - - -+
+         *     | entries added manually with                |
+         *     | addEntry() and addAllEntries()             |
+         *     + - - - - - - - - - - - - - - - - - - - - - -+
+         *     | APPEND_PLAYLIST                            |
+         *     | SHARE                                      |
+         *     | OPEN_IN_BROWSER                            |
+         *     | PLAY_WITH_KODI                             |
+         *     | MARK_AS_WATCHED                            |
+         *     | SHOW_CHANNEL_DETAILS                       |
+         *     + - - - - - - - - - - - - - - - - - - - - - -+
+         * </pre>
+         * Please note that some entries are not added depending on the user's preferences,
+         * the item's {@link StreamType} and the current player state.
+         *
+         * @param activity
+         * @param context
+         * @param fragment
+         * @param infoItem
+         * @param addDefaultEntriesAutomatically
+         *        whether default entries added with {@link #addDefaultBeginningEntries()}
+         *        and {@link #addDefaultEndEntries()} are added automatically when generating
+         *        the {@link InfoItemDialog}.
+         *        <br/>
+         *        Entries added with {@link #addEntry(StreamDialogDefaultEntry)} and
+         *        {@link #addAllEntries(StreamDialogDefaultEntry...)} are added in between.
+         */
+        public Builder(@NonNull final Activity activity,
+                       @NonNull final Context context,
+                       @NonNull final Fragment fragment,
+                       @NonNull final StreamInfoItem infoItem,
+                       final boolean addDefaultEntriesAutomatically) {
+            this.activity = activity;
+            this.context = context;
+            this.fragment = fragment;
+            this.infoItem = infoItem;
+            this.addDefaultEntriesAutomatically = addDefaultEntriesAutomatically;
+            if (addDefaultEntriesAutomatically) {
+                addDefaultBeginningEntries();
+            }
+        }
+
+        public void addEntry(@NonNull final StreamDialogDefaultEntry entry) {
```

Comment:
This is a `Builder`, so I think all of its function should be made so that they `return this`, so that method calls can be chained

Reply:
I thought about this as well. Returning `this` would be the "proper" implementation of a Builder.
I came to the conclusion that using chained calls makes the code hard to understand while refactoring the code (e.g. see the parts where actions are set `LocalPlaylistFragment` and `StatisticsPlaylistFragment`). For that reason, I did not return `this`. I'd refrain from using chained method calls, but I can still return `this` if you want.

Reply:
Ok, no, no need

Reply:
I find Stypox point still valid:

I think this:
```java
        // set custom actions; all entries modified here have already been added within the builder
        dialogBuilder
                .setAction(
                        StreamDialogDefaultEntry.START_HERE_ON_BACKGROUND,
                        (f, i) -> NavigationHelper.playOnBackgroundPlayer(
                                context, getPlayQueueStartingAt(item), true));
                .setAction(
                        StreamDialogDefaultEntry.SET_AS_PLAYLIST_THUMBNAIL,
                        (f, i) ->
                                changeThumbnailUrl(item.getStreamEntity().getThumbnailUrl()));
                .setAction(
                        StreamDialogDefaultEntry.DELETE,
                        (f, i) -> deleteItem(item));
```
is way more readable than
https://github.com/TeamNewPipe/NewPipe/blob/dbc7218501ea5acaf4f1436b2174f7cc5d1e11c6/app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java#L759-L767

because less code to read.

PS: Maybe it should also be called ``registerAction``